fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/AbstractStorageAdapter.java

```diff
@@ -0,0 +1,289 @@
+ * <p>
+ * This class provides the common infrastructure for managing HNSW graph data within {@link Subspace}.
+ * It handles the configuration, node creation, and listener management, while delegating the actual
+ * storage-specific read and write operations to concrete subclasses through the {@code fetchNodeInternal}
+ * and {@code writeNodeInternal} abstract methods.
+ *
+ * @param <N> the type of {@link NodeReference} used to reference nodes in the graph
+ */
+abstract class AbstractStorageAdapter<N extends NodeReference> implements StorageAdapter<N> {
+    @Nonnull
+    private static final Logger logger = LoggerFactory.getLogger(AbstractStorageAdapter.class);
+
+    @Nonnull
+    private final HNSW.Config config;
+    @Nonnull
+    private final NodeFactory<N> nodeFactory;
+    @Nonnull
+    private final Subspace subspace;
+    @Nonnull
+    private final OnWriteListener onWriteListener;
+    @Nonnull
+    private final OnReadListener onReadListener;
+
+    private final Subspace dataSubspace;
+
+    /**
+     * Constructs a new {@code AbstractStorageAdapter}.
+     * <p>
+     * This constructor initializes the adapter with the necessary configuration,
+     * factories, and listeners for managing an HNSW graph. It also sets up a
+     * dedicated data subspace within the provided main subspace for storing node data.
+     *
+     * @param config the HNSW graph configuration
+     * @param nodeFactory the factory to create new nodes of type {@code <N>}
+     * @param subspace the primary subspace for storing all graph-related data
+     * @param onWriteListener the listener to be called on write operations
+     * @param onReadListener the listener to be called on read operations
+     */
+    protected AbstractStorageAdapter(@Nonnull final HNSW.Config config, @Nonnull final NodeFactory<N> nodeFactory,
+                                     @Nonnull final Subspace subspace,
+                                     @Nonnull final OnWriteListener onWriteListener,
+                                     @Nonnull final OnReadListener onReadListener) {
+        this.config = config;
+        this.nodeFactory = nodeFactory;
+        this.subspace = subspace;
+        this.onWriteListener = onWriteListener;
+        this.onReadListener = onReadListener;
+        this.dataSubspace = subspace.subspace(Tuple.from(SUBSPACE_PREFIX_DATA));
+    }
+
+    /**
+     * Returns the configuration used to build and search this HNSW graph.
+     *
+     * @return the current {@link HNSW.Config} object, never {@code null}.
+     */
+    @Override
+    @Nonnull
+    public HNSW.Config getConfig() {
+        return config;
+    }
+
+    /**
+     * Gets the factory responsible for creating new nodes.
+     * <p>
+     * This factory is used to instantiate nodes of the generic type {@code N}
+     * for the current context. The {@code @Nonnull} annotation guarantees that
+     * this method will never return {@code null}.
```

Comment:
Some of the Javadocs here feel kind of wordy. I guess I don't fundamentally object, but it does already say `non-null` in the `@return`, too.